### PR TITLE
feat: add auto-fix for f-string logging statements

### LINKED ISSUE #19302: Add Autofix for f-string formatting G004 rule
> The current rule we need the most in microsoft/onnxruntime is G004. Because everything else can be turned into an f-string first. 

 _Originally posted by @justinchuby in [#15555](https://github.com/astral-sh/ruff/issues/15555#issuecomment-2598889866)_

### PR DESCRIPTION
**Pull request by hamirmahal**
Closes #19302

## Summary
This adds an auto-fix for `Logging statement uses f-string` Ruff G004, so users don't have to resolve it manually.

## Test Plan
I ran the auto-fixes on a Python file locally and and it worked as expected.

### CODE REVIEW

**Comment by MeGaGiGaGon** [comment]
How does this handle arbitrary format specs? Reading the code it looks to me like they all get converted to `%s`, which is definitely not ideal, and should at the very least make the fix unsafe/not offer a fix at all. As an example, what does the fix do on this code?
```py
import logging
logging.info(f"{x:arbitrary data in format spec} {x:{'nested format spec'}}")
```

**Comment by ntBre** [comment]
Thanks for working on this! I think @MeGaGiGaGon raises some good points. I think it would make sense to bail out and not offer a fix if we see any complicated format specs, especially for a first iteration. We could keep the `%d` and `%f` handling that it looks like you already worked on, but I'd even be happy with only handling `{var}` -> `%s` in a first version (omitting format specs entirely).

I can try to give this a review next week if you resolve the conflicts :)

**Comment by hirehamir** [comment]
You're welcome! That works for me. I pulled the latest changes and fixed the merge conflicts.

File: crates/ruff_linter/src/rules/flake8_logging_format/violations.rs [hunk 1]
```diff
@@ -327,6 +327,11 @@ impl Violation for LoggingStringConcat {
 pub(crate) struct LoggingFString;
 
 impl Violation for LoggingFString {
+    const FIX_AVAILABILITY: crate::FixAvailability = crate::FixAvailability::Sometimes;
+    fn fix_title(&self) -> Option<String> {
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/violations.rs]: tiny nit: We usually put this below `message`.

I also think it might be nice to wrap % in backticks:


```suggestion
    fn fix_title(&self) -> Option<String> {
        Some("Convert to lazy `%` formatting".to_string())
    }
```

This message feels very slightly misleading to me (it makes me think the fix is something like `"%s" % var` with the user adding a `%`), but I'm also not coming up with anything better, and they do need to add some `%`s themselves, in the quotes.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 2]
```diff
@@ -55,8 +56,65 @@ fn check_msg(checker: &Checker, msg: &Expr) {
             _ => {}
         },
         // Check for f-strings.
-        Expr::FString(_) => {
-            checker.report_diagnostic_if_enabled(LoggingFString, msg.range());
+        Expr::FString(f_string) => {
+            if checker.is_rule_enabled(Rule::LoggingFString) {
+                let mut format_string = String::new();
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: I think it would be nice to factor everything in the `is_rule_enabled` check into a function. We're getting pretty indented here.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 3]
```diff
@@ -55,8 +56,65 @@ fn check_msg(checker: &Checker, msg: &Expr) {
...
+                        "\"{}\"{}{}",
+                        format_string,
+                        if args.is_empty() {
+                            String::new()
+                        } else {
+                            String::from(", ")
+                        },
+                        args.join(", ")
+                    );
+                    let fix = Fix::safe_edit(Edit::range_replacement(replacement, msg.range()));
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: I believe that it's a breaking change to add a safe fix on stable: [docs](https://docs.astral.sh/ruff/versioning/#version-changes). We should gate this behind a preview check like these:

https://github.com/astral-sh/ruff/blob/cabdd969ec3b7807a777ac87de94771a67525777/crates/ruff_linter/src/preview.rs#L204-L207

You can do something like this:

```rust
let mut diagnostic = checker.report_diagnostic(LoggingFString, msg.range());

if !is_f_string_logging_fix_enabled(checker.settings()) {
    return
}

// fix code here
```

`report_diagnostic` ensures the diagnostic gets reported when the guard it returns is dropped, so if you return early without setting a fix, everything should work out.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 4]
```diff
@@ -55,8 +56,65 @@ fn check_msg(checker: &Checker, msg: &Expr) {
...
+
+                            InterpolatedStringElement::Interpolation(interpolated) => {
+                                if interpolated.format_spec.is_some()
+                                    || !matches!(
+                                        interpolated.conversion,
+                                        ruff_python_ast::ConversionFlag::None
+                                    )
+                                {
+                                    simple = false;
+                                    break;
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: I think this would also be helped by my suggestion to create the diagnostic early and return if we don't want a fix. Instead of setting a flag and breaking here, we can just return.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 5]
```diff
@@ -55,8 +56,65 @@ fn check_msg(checker: &Checker, msg: &Expr) {
...
+                    let replacement = format!(
+                        "\"{}\"{}{}",
+                        format_string,
+                        if args.is_empty() {
+                            String::new()
+                        } else {
+                            String::from(", ")
+                        },
+                        args.join(", ")
+                    );
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: nit: We usually try to use named arguments to `format!`. Something like this maybe:


```suggestion
                    let replacement = format!(
                        "\"{format_string}\"{sep}{args}",
                        sep = if args.is_empty() {
                            ""
                        } else {
                            ", "
                        },
                        args = args.join(", ")
                    );
```

I think you can also use string literals instead of converting to `String`s.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs (hunk 3 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: How does this fix handle an f-string with `%` formatting directives in it? I'm not sure if we have any utility functions for detecting that, but I think that would cause problems here. For example:

```python
import logging

x = 1
logging.error(f"{x} -> %s", x)
```

I think the current fix will overwrite the whole `logging.error` range. I think the logging functions also accept `**kwargs` as described on [`Logger.debug`](https://docs.python.org/3/library/logging.html#logging.Logger.debug), so we may need to be more careful about existing arguments in general.

**Review by ntBre — COMMENTED** [review]
Thanks! I had some organizational suggestions and also some edge cases I think we'll need to handle.

**Comment by hirehamir** [comment]
You're welcome! Weekends are usually when I have more availability; I'll see if I can commit your suggestions now, though.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs (hunk 3 above)
Comment (hirehamir) [reply crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: Hmm... I think the code in its current state does this.

```python
import logging

x = 1
logging.error(f"{x} -> %s", x) # original
logging.error("%s -> %s", x, x) # auto-fixed
```

I'm seeing the same output for both the original and fixed line, so I think this is valid? I could be missing something.

```
ERROR:root:1 -> 1
ERROR:root:1 -> 1
```

Comment (hirehamir) [reply crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: When I tried to [gate the f-string formatting fix behind a preview check](https://github.com/astral-sh/ruff/commit/07990a73916030c077b279b61ad473dc491630b4), that caused CI to fail.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 6]
```diff
@@ -11,6 +12,55 @@ use crate::rules::flake8_logging_format::violations::{
 };
 use crate::{Edit, Fix};
 
+fn handle_logging_f_string(checker: &Checker, msg: &Expr, f_string: &ast::ExprFString) {
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: I think we could report the diagnostic here at the top and then immediately return if preview is not enabled:

```rust
let mut diagnostic = checker.report_diagnostic(LoggingFString, msg.range());

if !is_fix_f_string_logging_enabled(checker.settings()) {
    return;
}
```

The diagnostic gets emitted when the guard returned by `report_diagnostic` is dropped, so we can do neat stuff like this. Later in the function we can just attach a fix to `diagnostic`, if needed.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 7]
```diff
@@ -11,6 +12,55 @@ use crate::rules::flake8_logging_format::violations::{
...
+                    format_string.push_str(lit.value.as_ref());
+                }
+                InterpolatedStringElement::Interpolation(interpolated) => {
+                    if interpolated.format_spec.is_some()
+                        || !matches!(
+                            interpolated.conversion,
+                            ruff_python_ast::ConversionFlag::None
+                        )
+                    {
+                        checker.report_diagnostic(LoggingFString, msg.range());
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: Yeah, then we can also drop these additional `report_diagnostic` calls and just return here, with the change above.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 8]
```diff
@@ -11,6 +12,55 @@ use crate::rules::flake8_logging_format::violations::{
...
+                            checker.report_diagnostic(LoggingFString, msg.range());
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if args.is_empty() {
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs:82]: This can also be an early return:

```rust
if args.is_empty() {
    return;
}
```

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 9]
```diff
@@ -11,6 +12,55 @@ use crate::rules::flake8_logging_format::violations::{
...
+        checker.report_diagnostic(LoggingFString, msg.range());
+    } else {
+        let replacement = format!(
+            "\"{format_string}\"{sep}{args}",
+            sep = if args.is_empty() { "" } else { ", " },
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: We already checked `args.is_empty()`, so this is always a comma.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 10]
```diff
@@ -55,8 +105,10 @@ fn check_msg(checker: &Checker, msg: &Expr) {
             _ => {}
         },
         // Check for f-strings.
-        Expr::FString(_) => {
-            checker.report_diagnostic_if_enabled(LoggingFString, msg.range());
+        Expr::FString(f_string) => {
+            if checker.is_rule_enabled(Rule::LoggingFString) {
+                handle_logging_f_string(checker, msg, f_string);
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: nit: I think we can just call the function `logging_f_string` to match the rule name.

We could also keep using `report_diagnostic_if_enabled` inside of the function, if we wanted. If it returns `None`, we'd just early return at the top of the function. This is fine too, though.

File: crates/ruff_linter/src/rules/flake8_logging_format/snapshots/ruff_linter__rules__flake8_logging_format__tests__G004.py.snap [hunk 11]
```diff
@@ -37,21 +70,209 @@ G004.py:10:26: G004 Logging statement uses f-string
...
+   |               ^^^^^^^^^^^^ G004
+   |
+   = help: Convert to lazy `%` formatting
+
+ℹ Safe fix
+63 63 | import logging
+64 64 | 
+65 65 | x = 1
+66    |-logging.error(f"{x} -> %s", x)
+   66 |+logging.error("%s -> %s", x, x)
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/snapshots/ruff_linter__rules__flake8_logging_format__tests__G004.py.snap]: We might want an example with a different variable name, just to make sure which `x` is which. That's my bad for the initial suggestion. Maybe something like this:

```py
y = 1
logging.error(f"{y} -> %s", x)
```

I just want to make sure it comes out as `logging.error("%s -> %s", y, x)` and not `x, y`. Hopefully that makes sense. We might also want to test the opposite order to be extra sure. I wouldn't really mind bailing out if the format string contains an existing `%` either, just to avoid an erroneous fix.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 12]
```diff
@@ -11,6 +12,55 @@ use crate::rules::flake8_logging_format::violations::{
...
+                }
+            }
+        }
+    }
+
+    if args.is_empty() {
+        checker.report_diagnostic(LoggingFString, msg.range());
+    } else {
+        let replacement = format!(
+            "\"{format_string}\"{sep}{args}",
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: I think we should try to reuse the f-string's quotes (at least the first element's if it's concatenated) here instead of hard-coding double quotes. You may need to add a `string_flags` method or something on `FStringPart`. I forget how we've done this before, but this should be a good start:

https://github.com/astral-sh/ruff/blob/1070e8cbf9211028a1add17d618ee7d7e78062e9/crates/ruff_python_ast/src/nodes.rs#L571-L576

Unfortunately `quote_style` isn't quite enough because it doesn't capture whether it's a single- or triple-quoted string. We probably need something like this if it doesn't already exist and I'm just overlooking it:

```rust
    pub fn quote_str(&self) -> &str {
        match self {
            Self::Literal(string_literal) => string_literal.flags.quote_str(),
            Self::FString(f_string) => f_string.flags.quote_str(),
        }
    }
```

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs (hunk 6 above)
Comment (hirehamir) [reply crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: Oh, that makes sense. Thanks for the suggestions.

**Comment by hamirmahal** [comment]
You're welcome! And no worries. I'll keep an eye out for any other changes I should make.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 13]
```diff
@@ -1,15 +1,109 @@
...
 
 use crate::checkers::ast::Checker;
+use crate::preview::is_fix_f_string_logging_enabled;
 use crate::registry::Rule;
 use crate::rules::flake8_logging_format::violations::{
     LoggingExcInfo, LoggingExtraAttrClash, LoggingFString, LoggingPercentFormat,
     LoggingRedundantExcInfo, LoggingStringConcat, LoggingStringFormat, LoggingWarn,
 };
 use crate::{Edit, Fix};
+fn logging_f_string(
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs:16]:
```suggestion

fn logging_f_string(
```

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 14]
```diff
@@ -1,15 +1,109 @@
...
+        .filter_map(|arg| match arg {
+            Expr::Name(name) => Some(name.id.to_string()),
+            _ => None,
+        })
+        .collect();
+
+    // Combine collected interpolation names with existing trailing positional names in
+    // that order so that placeholders map to arguments correctly. Do not deduplicate;
+    // duplicates are valid when the same name appears multiple times.
+    let mut result_args = args.clone();
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: Do we need to clone here? I think `args` is still a mutable `Vec` from above, so we could just extend it directly, unless I'm missing another binding that shadows line 31.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 15]
```diff
@@ -1,15 +1,109 @@
...
+
+    // Try to reuse the first part's quote style when building the replacement.
+    // Default to double quotes if we can't determine it.
+    let quote_str = f_string
+        .value
+        .f_strings()
+        .next()
+        .map(|f| f.flags.quote_str())
+        .unwrap_or("\"")
+        .to_string();
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: Do we need this to be a `String` or could we leave it as a `&str`? I could be missing it, but I'm not immediately seeing  a need for an owned string.

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 16]
```diff
@@ -1,15 +1,109 @@
...
+        return;
+    }
+
+    // Determine names of existing trailing positional args (after the `msg` argument).
+    let existing_names: Vec<String> = arguments
+        .args
+        .iter()
+        .skip(msg_pos + 1)
+        .filter_map(|arg| match arg {
+            Expr::Name(name) => Some(name.id.to_string()),
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs]: I _think_ we could drop the `to_string` calls on everything we put into `args` and `existing_names`. It looks to me like it would be fine to have a `Vec<&str>`, but it's not a big deal if it doesn't work out. Just a bit more efficient not to allocate `String`s if we can avoid it.

File: crates/ruff_linter/src/rules/flake8_logging_format/mod.rs [hunk 17]
```diff
@@ -48,4 +48,43 @@ mod tests {
         assert_diagnostics!(snapshot, diagnostics);
         Ok(())
     }
+
+    #[test]
+    fn g004_fix_gated_by_preview_flag() -> Result<()> {
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/mod.rs]: What we usually do for preview tests is create a copy of the `rules` function above this called `preview_rules` and run all of the same `test_case`s through both functions. That's a good way to compare preview and stable behavior across all the snapshots. Or maybe not all the files, since that would be a lot of snapshots :) Here's an example from a different rule with only a subset of the snapshots:

https://github.com/astral-sh/ruff/blob/39f7a876d6aceedc3db4e3d63250d78bbb830fc2/crates/ruff_linter/src/rules/flake8_comprehensions/mod.rs#L54-L58

I was playing with this branch locally and came up with this:

```rust
    #[test_case(Rule::LoggingFString, Path::new("G004.py"))]
    #[test_case(Rule::LoggingFString, Path::new("G004_arg_order.py"))]
    fn preview_rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!(
            "preview__{}_{}",
            rule_code.noqa_code(),
            path.to_string_lossy()
        );
        let diagnostics = test_path(
            Path::new("flake8_logging_format").join(path).as_path(),
            &settings::LinterSettings {
                logger_objects: vec!["logging_setup.logger".to_string()],
                preview: PreviewMode::Enabled,
                ..settings::LinterSettings::for_rule(rule_code)
            },
        )?;
        assert_diagnostics!(snapshot, diagnostics);
        Ok(())
    }
```

File: crates/ruff_linter/src/rules/flake8_logging_format/mod.rs [hunk 18]
```diff
@@ -48,4 +48,43 @@ mod tests {
...
+                .join(Path::new("G004_arg_order.py"))
+                .as_path(),
+            &LinterSettings {
+                preview: crate::settings::types::PreviewMode::Disabled,
+                ..LinterSettings::for_rules(vec![Rule::LoggingFString])
+            },
+        )?;
+
+        // When preview is disabled, diagnostics should be emitted but no fix should be attached.
+        assert!(diagnostics.iter().all(|d| d.fix().is_none()));
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/mod.rs]: See my comment above, it's also good to see the snapshots for this. I was wondering where the other snapshot file was!

File: crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs [hunk 19]
```diff
@@ -1,15 +1,109 @@
...
+    for f in f_string.value.f_strings() {
+        for element in &f.elements {
+            match element {
+                InterpolatedStringElement::Literal(lit) => {
+                    // If the literal text contains a '%' placeholder, bail out: mixing
+                    // f-string interpolation with '%' placeholders is ambiguous for our
+                    // automatic conversion, so don't offer a fix for this case.
+                    if lit.value.as_ref().contains('%') {
+                        return;
+                    }
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/rules/logging_call.rs:58]: I think this check is slightly too conservative  because `%%` escapes the `%` and would be okay, but I think that's fine. It would complicate our analysis substantially because we'd have to make sure not to add a `%s` next to an existing `%`. I spent a little while playing with this and think this simple check is probably best for now.

File: crates/ruff_linter/src/rules/flake8_logging_format/snapshots/ruff_linter__rules__flake8_logging_format__tests__G004_arg_order.py.snap [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+10 | logger.error(f"{Y} -> %s", X)
+   |
+help: Convert to lazy `%` formatting
+
+G004 Logging statement uses f-string
+  --> G004_arg_order.py:10:14
+   |
+ 8 | Y = 2
+ 9 | logger.error(f"{X} -> %s", Y)
+10 | logger.error(f"{Y} -> %s", X)
```
Comment (ntBre) [inline crates/ruff_linter/src/rules/flake8_logging_format/snapshots/ruff_linter__rules__flake8_logging_format__tests__G004_arg_order.py.snap:20]: Ahhh, I forgot that we're already bailing out if an existing `%` is present, so we don't have to worry about these fixes. I think we should also bail out if there are any existing positional arguments since that can also be a problem. And then we can avoid the `existing_names` code above.

To reach that point of the code, the user would have to have existing positional arguments but no `%` placeholders in the string, which either indicates a mistake, or even worse, an f-string like this:

```py
placeholder = "%s"
logging.error(f"{placeholder}", 123)
```

which we also shouldn't try to fix.

**Review by ntBre — COMMENTED** [review]
Thank you again, this is very nearly ready. I mostly had very small performance-related nits and a suggestion on test placement. There's also one last edge case I came up with, but I think we can mostly delete code to solve it, which is always nice.

**Comment by hamirmahal** [comment]
You're welcome! I'll work on the changes right now.

**Review by ntBre — APPROVED** [review]
Excellent, thanks again!

I went through the ecosystem changes, and they look good too. We don't show the suggested fix, but they all look like simple cases that I'd expect the fix to handle.

**Comment by hirehamir** [comment]
You're welcome! Thanks for the suggestions and feedback.
